File: src/panoptikon/db/pql/filters/sortable/text_embeddings.py
import base64
import io
import logging
import time
from typing import List, Literal, Optional

import numpy as np
from pydantic import BaseModel, Field, PrivateAttr
from sqlalchemy import and_, func, literal
from sqlalchemy.sql.expression import CTE, select

from inferio.impl.utils import deserialize_array
from panoptikon.db.pql.filters.sortable.item_similarity import SourceArgs
from panoptikon.db.pql.filters.sortable.sortable_filter import SortableFilter
from panoptikon.db.pql.filters.sortable.utils import extract_embeddings
from panoptikon.db.pql.types import (
    OrderTypeNN,
    QueryState,
    get_order_by_field,
    get_order_direction_field,
    get_std_cols,
    get_std_group_by,
)
from panoptikon.db.utils import serialize_f32

logger = logging.getLogger(__name__)


class EmbedArgs(BaseModel):
    cache_key: str = Field(
        default="search",
        title="Cache Key",
        description="The cache key to use for the inference *model*",
    )
    lru_size: int = Field(
        default=1,
        ge=1,
        title="LRU Cache Size",
        description="The size of the LRU cache to use for the inference *model*",
    )
    ttl_seconds: int = Field(
        default=60,
        title="TTL Seconds",
        description="The time-to-live in seconds for the inference *model* to be kept in memory",
    )


class SemanticTextArgs(BaseModel):
    query: str = Field(
        ...,
        title="Query",
        description="Semantic query to match against the text",
    )

    _embedding: Optional[bytes] = PrivateAttr(None)

    model: str = Field(
        title="The text embedding model to use",
        description="""
The text embedding model to use for the semantic search.
Will search embeddings produced by this model.
""",
    )
    distance_aggregation: Literal["MIN", "MAX", "AVG"] = Field(
        default="MIN",
        description="The method to aggregate distances when an item has multiple embeddings. Default is MIN.",
    )
    embed: EmbedArgs = Field(
        default_factory=EmbedArgs,
        title="Embed The Query",
        description="""
Embed the query using the model already specified in `model`.
This is useful when the query is a string and needs to be converted to an embedding.

If this is not present, the query is assumed to be an embedding already.
In that case, it must be a base64 encoded string of a numpy array.
        """,
    )
    src_text: Optional[SourceArgs] = Field(
        default=None,
        description="""
Filters and options to apply on source text that the embeddings are derived from.
""",
    )


class SemanticTextSearch(SortableFilter):
    order_by: bool = get_order_by_field(True)
    direction: OrderTypeNN = get_order_direction_field("asc")
    text_embeddings: SemanticTextArgs = Field(
        ...,
        title="Search Text Embeddings",
        description="""
Search for text using semantic search on text embeddings.
""",
    )

    def _validate(self):
        if len(self.text_embeddings.query.strip()) == 0:
            return self.set_validated(False)

        if self.text_embeddings.embed:
            self.text_embeddings._embedding = get_embed(
                self.text_embeddings.query,
                self.text_embeddings.model,
                self.text_embeddings.embed,
            )

        else:
            self.text_embeddings._embedding = extract_embeddings(
                self.text_embeddings.query
            )

        return self.set_validated(True)

    def build_query(self, context: CTE, state: QueryState) -> CTE:
        self.raise_if_not_validated()
        from panoptikon.db.pql.tables import (
            embeddings,
            extracted_text,
            item_data,
            setters,
        )

        args = self.text_embeddings
        criteria = []
        text_data = item_data.alias("text_data")
        text_setters = setters.alias("text_setters")
        vec_data = item_data.alias("vec_data")
        vec_setters = setters.alias("vec_setters")
        if args.src_text:
            if args.src_text.min_length:
                criteria.append(
                    extracted_text.c.text_length >= args.src_text.min_length
                )
            if args.src_text.max_length:
                criteria.append(
                    extracted_text.c.text_length <= args.src_text.max_length
                )
            if args.src_text.setters:
                criteria.append(text_setters.c.name.in_(args.src_text.setters))
            if args.src_text.languages:
                criteria.append(
                    extracted_text.c.language.in_(args.src_text.languages)
                )
            if args.src_text.min_language_confidence:
                criteria.append(
                    extracted_text.c.language_confidence
                    >= args.src_text.min_language_confidence
                )
            if args.src_text.min_confidence:
                criteria.append(
                    extracted_text.c.confidence >= args.src_text.min_confidence
                )

        vec_distance = func.vec_distance_L2(
            embeddings.c.embedding, literal(args._embedding)
        )
        if args.distance_aggregation == "MAX":
            rank_column = func.max(vec_distance)
        elif args.distance_aggregation == "AVG":
            rank_column = func.avg(vec_distance)
        elif args.distance_aggregation == "MIN":
            rank_column = func.min(vec_distance)
        else:
            raise ValueError(
                f"Invalid distance aggregation method: {args.distance_aggregation}"
            )
        if args.src_text:
            conf_weight_clause = func.pow(
                func.coalesce(extracted_text.c.confidence, 1),
                args.src_text.confidence_weight,
            )
            lang_conf_weight_clause = func.pow(
                func.coalesce(extracted_text.c.language_confidence, 1),
                args.src_text.language_confidence_weight,
            )
            if (
                args.src_text.confidence_weight != 0
                and args.src_text.language_confidence_weight != 0
            ):
                weights = conf_weight_clause * lang_conf_weight_clause
                rank_column = func.sum(vec_distance * weights) / func.sum(
                    weights
                )
            elif args.src_text.confidence_weight != 0:
                rank_column = func.sum(
                    vec_distance * conf_weight_clause
                ) / func.sum(conf_weight_clause)
            elif args.src_text.language_confidence_weight != 0:
                rank_column = func.sum(
                    vec_distance * lang_conf_weight_clause
                ) / func.sum(lang_conf_weight_clause)

        if state.item_data_query and state.entity == "text":
            return self.wrap_query(
                select(
                    *get_std_cols(context, state),
                    self.derive_rank_column(rank_column),
                )
                .join(
                    text_data,
                    (text_data.c.id == context.c.data_id),
                )
                .join(
                    text_setters,
                    text_setters.c.id == text_data.c.setter_id,
                )
                .join(
                    extracted_text,
                    context.c.data_id == extracted_text.c.id,
                )
                .join(
                    vec_data,
                    vec_data.c.source_id == extracted_text.c.id,
                )
                .join(
                    vec_setters,
                    (vec_setters.c.id == vec_data.c.setter_id)
                    & (vec_setters.c.name == args.model),
                )
                .join(
                    embeddings,
                    embeddings.c.id == vec_data.c.id,
                )
                .where(and_(*criteria))
                .group_by(*get_std_group_by(context, state)),
                context,
                state,
            )
        embeddings_query = (
            select(
                *get_std_cols(context, state),
                self.derive_rank_column(rank_column),
            )
            .join(
                vec_data,
                vec_data.c.item_id == context.c.item_id,
            )
            .join(
                vec_setters,
                (vec_setters.c.id == vec_data.c.setter_id)
                & (vec_setters.c.name == args.model),
            )
            .join(
                embeddings,
                embeddings.c.id == vec_data.c.id,
            )
            .where(and_(*criteria))
            .group_by(*get_std_group_by(context, state))
        )
        if len(criteria) > 0:
            embeddings_query = (
                embeddings_query.join(
                    text_data,
                    text_data.c.id == vec_data.c.source_id,
                )
                .join(
                    text_setters,
                    text_setters.c.id == text_data.c.setter_id,
                )
                .join(
                    extracted_text,
                    text_data.c.id == extracted_text.c.id,
                )
            )

        return self.wrap_query(
            embeddings_query,
            context,
            state,
        )


last_embedded_text: str | None = None
last_embedded_text_embed: bytes | None = None
last_used_model: str | None = None


def get_embed(
    text: str,
    model_name: str,
    cache_args: EmbedArgs,
) -> bytes:

    global last_embedded_text, last_embedded_text_embed, last_used_model
    if (
        text == last_embedded_text
        and model_name == last_used_model
        and last_embedded_text_embed is not None
    ):
        return last_embedded_text_embed

    from panoptikon.data_extractors.models import run_batch_inference

    logger.debug(f"Getting embedding for text: {text}")
    start_time = time.time()
    embed_bytes: bytes = run_batch_inference(
        model_name,
        cache_args.cache_key,
        cache_args.lru_size,
        cache_args.ttl_seconds,
        [({"text": text, "task": "s2s"}, None)],
    )[0]
    deserialized_embedding = deserialize_array(embed_bytes)
    if isinstance(deserialized_embedding[0], np.ndarray):
        text_embed = deserialized_embedding[0]
    else:
        text_embed = deserialized_embedding
    # Set as persistent so that the model is not reloaded every time the function is called
    last_embedded_text = text
    last_used_model = model_name
    embed_list = text_embed.tolist()
    assert isinstance(embed_list, list), "Expected a list"
    last_embedded_text_embed = serialize_f32(embed_list)
    logger.debug(
        f"Embedding generation took {time.time() - start_time} seconds"
    )
    return last_embedded_text_embed

File: gateway/src/pql/builder/filters/text_embeddings.rs
use sea_query::{Alias, Cond, Expr, ExprTrait, Func, JoinType};
use serde::{Deserialize, Serialize};
use utoipa::ToSchema;

use crate::pql::model::{EntityType, OrderDirection, SortableOptions};
use crate::pql::preprocess::PqlError;

use super::super::{
    BaseTable, CteRef, Embeddings, ExtraColumn, ExtractedText, ItemData, JoinedTables,
    OrderByFilter, QueryState, Setters, add_rank_column_expr, apply_group_by, apply_sort_bounds,
    get_std_group_by, select_std_from_cte, wrap_query,
};
use super::FilterCompiler;
use super::embedding_types::DistanceAggregation;
use super::item_similarity::SourceArgs;

#[derive(Debug, Clone, Serialize, Deserialize, ToSchema)]
pub(crate) struct EmbedArgs {
    /// Cache Key
    ///
    /// The cache key to use for the inference *model*
    #[serde(default = "default_cache_key")]
    pub cache_key: String,
    /// LRU Cache Size
    ///
    /// The size of the LRU cache to use for the inference *model*
    #[serde(default = "default_lru_size")]
    pub lru_size: i64,
    /// TTL Seconds
    ///
    /// The time-to-live in seconds for the inference *model* to be kept in memory
    #[serde(default = "default_ttl_seconds")]
    pub ttl_seconds: i64,
}

impl Default for EmbedArgs {
    fn default() -> Self {
        Self {
            cache_key: default_cache_key(),
            lru_size: default_lru_size(),
            ttl_seconds: default_ttl_seconds(),
        }
    }
}

#[derive(Debug, Clone, Serialize, Deserialize, ToSchema)]
pub(crate) struct SemanticTextArgs {
    /// Query
    ///
    /// Semantic query to match against the text
    pub query: String,
    #[serde(skip)]
    pub _embedding: Option<Vec<u8>>,
    /// The text embedding model to use
    ///
    /// The text embedding model to use for the semantic search.
    /// Will search embeddings produced by this model.
    pub model: String,
    /// The method to aggregate distances when an item has multiple embeddings. Default is MIN.
    #[serde(default)]
    pub distance_aggregation: DistanceAggregation,
    /// Embed The Query
    ///
    /// Embed the query using the model already specified in `model`.
    /// This is useful when the query is a string and needs to be converted to an embedding.
    ///
    /// If this is not present, the query is assumed to be an embedding already.
    /// In that case, it must be a base64 encoded string of a numpy array.
    #[serde(default = "default_embed_args")]
    pub embed: Option<EmbedArgs>,
    /// Filters and options to apply on source text that the embeddings are derived from.
    #[serde(default)]
    pub src_text: Option<SourceArgs>,
}

#[derive(Debug, Clone, Serialize, Deserialize, ToSchema)]
pub(crate) struct SemanticTextSearch {
    #[serde(flatten, default = "default_sort_asc")]
    pub sort: SortableOptions,
    /// Search Text Embeddings
    ///
    /// Search for text using semantic search on text embeddings.
    pub text_embeddings: SemanticTextArgs,
}

fn default_cache_key() -> String {
    "search".to_string()
}

fn default_lru_size() -> i64 {
    1
}

fn default_ttl_seconds() -> i64 {
    60
}

fn default_embed_args() -> Option<EmbedArgs> {
    Some(EmbedArgs::default())
}

// Used by serde default attribute.
#[allow(dead_code)]
fn default_sort_asc() -> SortableOptions {
    let mut options = SortableOptions::default();
    options.order_by = true;
    options.direction = OrderDirection::Asc;
    options.row_n_direction = OrderDirection::Asc;
    options
}

impl FilterCompiler for SemanticTextSearch {
    fn build(&self, context: &CteRef, state: &mut QueryState) -> Result<CteRef, PqlError> {
        let args = &self.text_embeddings;
        let embedding = args
            ._embedding
            .as_ref()
            .ok_or_else(|| PqlError::invalid("text_embeddings missing embedding bytes"))?;
        let cte_name = format!("n{}_SemanticTextSearch", state.cte_counter);

        let text_data = Alias::new("text_data");
        let text_setters = Alias::new("text_setters");
        let vec_data = Alias::new("vec_data");
        let vec_setters = Alias::new("vec_setters");

        let mut criteria = Vec::new();
        let mut weights_used = false;
        if let Some(src_text) = &args.src_text {
            if src_text.min_length > 0 {
                criteria.push(
                    Expr::col((ExtractedText::Table, ExtractedText::TextLength))
                        .gte(src_text.min_length),
                );
            }
            if let Some(max_length) = src_text.max_length {
                if max_length > 0 {
                    criteria.push(
                        Expr::col((ExtractedText::Table, ExtractedText::TextLength))
                            .lte(max_length),
                    );
                }
            }
            if !src_text.setters.is_empty() {
                let setters = src_text
                    .setters
                    .iter()
                    .cloned()
                    .map(Expr::val)
                    .collect::<Vec<_>>();
                criteria.push(Expr::col((text_setters.clone(), Setters::Name)).is_in(setters));
            }
            if let Some(languages) = &src_text.languages {
                if !languages.is_empty() {
                    let values = languages.iter().cloned().map(Expr::val).collect::<Vec<_>>();
                    criteria.push(
                        Expr::col((ExtractedText::Table, ExtractedText::Language)).is_in(values),
                    );
                }
            }
            if src_text.min_language_confidence > 0.0 {
                criteria.push(
                    Expr::col((ExtractedText::Table, ExtractedText::LanguageConfidence))
                        .gte(src_text.min_language_confidence),
                );
            }
            if let Some(min_confidence) = src_text.min_confidence {
                if min_confidence > 0.0 {
                    criteria.push(
                        Expr::col((ExtractedText::Table, ExtractedText::Confidence))
                            .gte(min_confidence),
                    );
                }
            }
            if src_text.confidence_weight != 0.0 || src_text.language_confidence_weight != 0.0 {
                weights_used = true;
            }
        }

        let vec_distance: Expr = Func::cust("vec_distance_L2")
            .args([
                Expr::col((Embeddings::Table, Embeddings::Embedding)),
                Expr::val(embedding.clone()),
            ])
            .into();
        let mut rank_column = match args.distance_aggregation {
            DistanceAggregation::Max => vec_distance.clone().max(),
            DistanceAggregation::Avg => vec_distance.clone().avg(),
            DistanceAggregation::Min => vec_distance.clone().min(),
        };

        if let Some(src_text) = &args.src_text {
            let conf_weight_clause: Expr = Func::cust("pow")
                .args([
                    Func::coalesce([
                        Expr::col((ExtractedText::Table, ExtractedText::Confidence)),
                        Expr::val(1),
                    ])
                    .into(),
                    Expr::val(src_text.confidence_weight),
                ])
                .into();
            let lang_conf_weight_clause: Expr = Func::cust("pow")
                .args([
                    Func::coalesce([
                        Expr::col((ExtractedText::Table, ExtractedText::LanguageConfidence)),
                        Expr::val(1),
                    ])
                    .into(),
                    Expr::val(src_text.language_confidence_weight),
                ])
                .into();
            if src_text.confidence_weight != 0.0 && src_text.language_confidence_weight != 0.0 {
                let weights = conf_weight_clause
                    .clone()
                    .mul(lang_conf_weight_clause.clone());
                rank_column = vec_distance
                    .clone()
                    .mul(weights.clone())
                    .sum()
                    .div(weights.sum());
            } else if src_text.confidence_weight != 0.0 {
                rank_column = vec_distance
                    .clone()
                    .mul(conf_weight_clause.clone())
                    .sum()
                    .div(conf_weight_clause.sum());
            } else if src_text.language_confidence_weight != 0.0 {
                rank_column = vec_distance
                    .clone()
                    .mul(lang_conf_weight_clause.clone())
                    .sum()
                    .div(lang_conf_weight_clause.sum());
            }
        }

        if state.item_data_query && matches!(state.entity, EntityType::Text) {
            let mut query = select_std_from_cte(context, state);
            query.join_as(
                JoinType::InnerJoin,
                ItemData::Table,
                text_data.clone(),
                Expr::col((text_data.clone(), ItemData::Id)).equals(context.column_ref("data_id")),
            );
            query.join_as(
                JoinType::InnerJoin,
                Setters::Table,
                text_setters.clone(),
                Expr::col((text_setters.clone(), Setters::Id))
                    .equals((text_data.clone(), ItemData::SetterId)),
            );
            query.join(
                JoinType::InnerJoin,
                ExtractedText::Table,
                Expr::col((ExtractedText::Table, ExtractedText::Id))
                    .equals(context.column_ref("data_id")),
            );
            query.join_as(
                JoinType::InnerJoin,
                ItemData::Table,
                vec_data.clone(),
                Expr::col((vec_data.clone(), ItemData::SourceId))
                    .equals((ExtractedText::Table, ExtractedText::Id)),
            );
            let vec_join = Cond::all()
                .add(
                    Expr::col((vec_setters.clone(), Setters::Id))
                        .equals((vec_data.clone(), ItemData::SetterId)),
                )
                .add(Expr::col((vec_setters.clone(), Setters::Name)).eq(args.model.clone()));
            query.join_as(
                JoinType::InnerJoin,
                Setters::Table,
                vec_setters.clone(),
                vec_join,
            );
            query.join(
                JoinType::InnerJoin,
                Embeddings::Table,
                Expr::col((Embeddings::Table, Embeddings::Id))
                    .equals((vec_data.clone(), ItemData::Id)),
            );

            for condition in &criteria {
                query.and_where(condition.clone());
            }

            apply_group_by(&mut query, get_std_group_by(context, state));
            if !state.is_count_query {
                add_rank_column_expr(&mut query, &self.sort, rank_column)?;
            }

            let (query, context_for_wrap) =
                apply_sort_bounds(state, query, context.clone(), &cte_name, &self.sort);

            let mut joined_tables = JoinedTables::default();
            joined_tables.mark(BaseTable::ItemData);
            joined_tables.mark(BaseTable::Setters);
            joined_tables.mark(BaseTable::ExtractedText);
            let cte = wrap_query(state, query, &context_for_wrap, cte_name, &joined_tables);
            state.cte_counter += 1;
            if !state.is_count_query {
                if let Some(alias) = &self.sort.select_as {
                    state.extra_columns.push(ExtraColumn {
                        column: "order_rank".to_string(),
                        cte: cte.clone(),
                        alias: alias.clone(),
                    });
                }
                if self.sort.order_by {
                    state.order_list.push(OrderByFilter {
                        cte: cte.clone(),
                        direction: self.sort.direction,
                        priority: self.sort.priority,
                        rrf: self.sort.rrf.clone(),
                    });
                }
            }
            return Ok(cte);
        }

        let mut query = select_std_from_cte(context, state);
        query.join_as(
            JoinType::InnerJoin,
            ItemData::Table,
            vec_data.clone(),
            Expr::col((vec_data.clone(), ItemData::ItemId)).equals(context.column_ref("item_id")),
        );
        let vec_join = Cond::all()
            .add(
                Expr::col((vec_setters.clone(), Setters::Id))
                    .equals((vec_data.clone(), ItemData::SetterId)),
            )
            .add(Expr::col((vec_setters.clone(), Setters::Name)).eq(args.model.clone()));
        query.join_as(
            JoinType::InnerJoin,
            Setters::Table,
            vec_setters.clone(),
            vec_join,
        );
        query.join(
            JoinType::InnerJoin,
            Embeddings::Table,
            Expr::col((Embeddings::Table, Embeddings::Id)).equals((vec_data.clone(), ItemData::Id)),
        );

        if !criteria.is_empty() || weights_used {
            query.join_as(
                JoinType::InnerJoin,
                ItemData::Table,
                text_data.clone(),
                Expr::col((text_data.clone(), ItemData::Id))
                    .equals((vec_data.clone(), ItemData::SourceId)),
            );
            query.join_as(
                JoinType::InnerJoin,
                Setters::Table,
                text_setters.clone(),
                Expr::col((text_setters.clone(), Setters::Id))
                    .equals((text_data.clone(), ItemData::SetterId)),
            );
            query.join(
                JoinType::InnerJoin,
                ExtractedText::Table,
                Expr::col((ExtractedText::Table, ExtractedText::Id))
                    .equals((text_data.clone(), ItemData::Id)),
            );
        }
        for condition in &criteria {
            query.and_where(condition.clone());
        }

        apply_group_by(&mut query, get_std_group_by(context, state));
        if !state.is_count_query {
            add_rank_column_expr(&mut query, &self.sort, rank_column)?;
        }

        let (query, context_for_wrap) =
            apply_sort_bounds(state, query, context.clone(), &cte_name, &self.sort);

        let mut joined_tables = JoinedTables::default();
        joined_tables.mark(BaseTable::ItemData);
        joined_tables.mark(BaseTable::Setters);
        if !criteria.is_empty() || weights_used {
            joined_tables.mark(BaseTable::ExtractedText);
        }
        let cte = wrap_query(state, query, &context_for_wrap, cte_name, &joined_tables);
        state.cte_counter += 1;
        if !state.is_count_query {
            if let Some(alias) = &self.sort.select_as {
                state.extra_columns.push(ExtraColumn {
                    column: "order_rank".to_string(),
                    cte: cte.clone(),
                    alias: alias.clone(),
                });
            }
            if self.sort.order_by {
                state.order_list.push(OrderByFilter {
                    cte: cte.clone(),
                    direction: self.sort.direction,
                    priority: self.sort.priority,
                    rrf: self.sort.rrf.clone(),
                });
            }
        }
        Ok(cte)
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::pql::model::{EntityType, QueryElement};
    use serde_json::json;

    use super::super::test_support::{
        build_base_state, build_begin_cte, render_filter_sql, run_full_pql_query,
    };

    #[test]
    fn semantic_text_builds_sql() {
        let mut filter: SemanticTextSearch = serde_json::from_value(json!({
            "text_embeddings": { "query": "hello", "model": "textembed/test" }
        }))
        .expect("semantic text filter");
        filter.text_embeddings._embedding = Some(vec![0, 0, 0, 0]);
        let mut state = build_base_state(EntityType::File, false);
        let context = build_begin_cte(&mut state);
        let sql = render_filter_sql(&filter, &mut state, &context);
        assert!(sql.contains("embeddings"));
    }

    #[tokio::test]
    async fn semantic_text_runs_full_query() {
        let mut filter: SemanticTextSearch = serde_json::from_value(json!({
            "text_embeddings": { "query": "hello", "model": "textembed/test" }
        }))
        .expect("semantic text filter");
        filter.text_embeddings._embedding = Some(vec![0, 0, 0, 0]);
        run_full_pql_query(QueryElement::SemanticTextSearch(filter), EntityType::File)
            .await
            .expect("semantic text query");
    }
}
